File: src/main/java/io/springreact/live/LiveComponentRegistry.java
package io.springreact.live;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.context.ApplicationContext;
import org.springframework.util.ClassUtils;

/**
 * Discovers all {@link LiveComponent}-annotated beans at startup and creates fresh,
 * fully dependency-injected instances on demand (one per mounted client component).
 */
public class LiveComponentRegistry {

    private final ApplicationContext context;
    private final Map<String, Class<?>> componentsByName = new HashMap<>();
    private final Map<Class<?>, LiveComponentDescriptor> descriptors = new ConcurrentHashMap<>();

    public LiveComponentRegistry(ApplicationContext context) {
        this.context = context;
        for (String beanName : context.getBeanNamesForAnnotation(LiveComponent.class)) {
            Class<?> type = context.getType(beanName);
            LiveComponent annotation = context.findAnnotationOnBean(beanName, LiveComponent.class);
            if (type != null && annotation != null) {
                componentsByName.put(annotation.value(), ClassUtils.getUserClass(type));
            }
        }
    }

    public boolean has(String name) {
        return componentsByName.containsKey(name);
    }

    public Set<String> names() {
        return componentsByName.keySet();
    }

    /** Create a new, autowired instance of the named component. */
    public Object create(String name) {
        Class<?> type = componentsByName.get(name);
        if (type == null) {
            throw new IllegalArgumentException("Unknown live component: '" + name + "'");
        }
        // createBean() always returns a fresh instance with full constructor/field
        // injection, regardless of the bean's declared scope.
        return context.getAutowireCapableBeanFactory().createBean(type);
    }

    public LiveComponentDescriptor descriptor(Object instance) {
        return descriptors.computeIfAbsent(ClassUtils.getUserClass(instance),
                LiveComponentDescriptor::of);
    }
}

File: src/main/kotlin/io/springreact/live/LiveComponentRegistry.kt
package io.springreact.live

import org.springframework.context.ApplicationContext
import org.springframework.util.ClassUtils
import org.springframework.util.ReflectionUtils
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap

/** Reflective metadata for one component class: its `@LiveAction`s and `@LiveState` fields. */
class LiveComponentDescriptor private constructor() {

    private val actions = HashMap<String, Method>()
    private val stateFields = ArrayList<Field>()
    private val paramFields = ArrayList<Pair<String, Field>>()

    fun action(name: String): Method? = actions[name]

    /** Route-param name → field, for `@LiveParam` fields. */
    fun paramFields(): List<Pair<String, Field>> = paramFields

    fun readState(instance: Any): Map<String, Any?> {
        val state = LinkedHashMap<String, Any?>()
        for (field in stateFields) state[field.name] = ReflectionUtils.getField(field, instance)
        return state
    }

    companion object {
        fun of(type: Class<*>): LiveComponentDescriptor {
            val d = LiveComponentDescriptor()
            ReflectionUtils.doWithMethods(type) { method ->
                method.getAnnotation(LiveAction::class.java)?.let { a ->
                    val name = a.value.ifEmpty { method.name }
                    method.isAccessible = true
                    d.actions[name] = method
                }
            }
            ReflectionUtils.doWithFields(type) { field ->
                if (field.isAnnotationPresent(LiveState::class.java)) {
                    field.isAccessible = true
                    d.stateFields.add(field)
                }
                field.getAnnotation(LiveParam::class.java)?.let { a ->
                    field.isAccessible = true
                    d.paramFields.add((a.value.ifEmpty { field.name }) to field)
                }
            }
            return d
        }
    }
}

/**
 * Discovers all `@LiveComponent` beans at startup and creates fresh, fully injected
 * instances on demand (one per mounted client component).
 */
class LiveComponentRegistry(private val context: ApplicationContext) {

    private val componentsByName = HashMap<String, Class<*>>()
    private val descriptors = ConcurrentHashMap<Class<*>, LiveComponentDescriptor>()

    init {
        for (beanName in context.getBeanNamesForAnnotation(LiveComponent::class.java)) {
            val type = context.getType(beanName)
            val annotation = context.findAnnotationOnBean(beanName, LiveComponent::class.java)
            if (type != null && annotation != null) {
                componentsByName[annotation.value] = ClassUtils.getUserClass(type)
            }
        }
    }

    fun has(name: String) = componentsByName.containsKey(name)

    fun create(name: String): Any {
        val type = componentsByName[name] ?: throw IllegalArgumentException("Unknown live component: '$name'")
        // Always a fresh, fully-wired instance regardless of declared scope.
        return context.autowireCapableBeanFactory.createBean(type)
    }

    fun descriptor(instance: Any): LiveComponentDescriptor =
        descriptors.computeIfAbsent(ClassUtils.getUserClass(instance)) { LiveComponentDescriptor.of(it) }
}
